flasksrc/session.py

```diff
@@ -607,6 +625,105 @@ def put(self, toynet_session_id, **kwargs):
             }, 200
 
 
+class ToyNetSessionByIdCreateRouterInterface(MethodResource):
+    @use_kwargs(ToyNetSessionByIdCreateRouterInterfacePutReq)
+    @marshal_with(ToyNetSessionByIdCreateRouterInterfacePutResp)
+    def put(self, toynet_session_id, **kwargs):
+        try:
+            req = ToyNetSessionByIdCreateRouterInterfacePutReq().load(kwargs)
+        except ValidationError as e:
+            abort(400, message='Invalid request: {}'.format(e))
+
+        # Separate validation from Marshmallow
+        if 'ip' not in req:
+            abort(400, message='Missing ip from req')
+        elif 'name' not in req:
+            abort(400, message='Missing name from req')
+
+        orig_topology = getTopologyFromDb(toynet_session_id)['topology']
+        root = ET.fromstring(orig_topology)
+
+        # Get the router to add the interface to, note that we do not validate
+        # that the IP address is "good" (eg: not a duplicate of another
+        # interface)
+        router = None
+        router_list = root.find('routerList').findall('router')
+
+        for device in router_list:
+            if device.get('name') == req['name']:
+                router = device
+
+        if router is None:
+            abort(400, message=f'Router {req["name"]} does not exist')
+
+        # Create XML elements and update topology
+        interface = ET.Element('intf')
+        interface.text = req['ip']
+        router.append(interface)
+        new_topology = ET.tostring(root).decode('utf-8')
+
+        # Send to mininet and update DB, these functions abort the REST call on
+        # failure
+        sendTopoToMininet(toynet_session_id, new_topology)
+        updateTopoInDb(toynet_session_id, new_topology)
+
+        return {
+            }, 200
+
+
+class ToyNetSessionByIdDeleteRouterInterface(MethodResource):
+    @use_kwargs(ToyNetSessionByIdDeleteRouterInterfacePutReq)
+    @marshal_with(ToyNetSessionByIdDeleteRouterInterfacePutResp)
+    def put(self, toynet_session_id, **kwargs):
+        try:
+            req = ToyNetSessionByIdDeleteRouterInterfacePutReq().load(kwargs)
+        except ValidationError as e:
+            abort(400, message='Invalid request: {}'.format(e))
+
+        # Separate validation from Marshmallow
+        if 'ip' not in req:
+            abort(400, message='Missing ip from req')
+        elif 'name' not in req:
+            abort(400, message='Missing name from req')
+
+        orig_topology = getTopologyFromDb(toynet_session_id)['topology']
+        root = ET.fromstring(orig_topology)
+
+        # Update the topology and get the interface index
+        intf_idx = None
+        for device in root.find('routerList'):
```

Comment:
There are enough indentations here making me dizzy that I would prefer if there's one layer of abstraction via a helper function. Thoughts?

Reply:
I don't think a helper function is very useful here since the logic isn't repeated elsewhere. I could not find a good way to get the specific device/interface elements from the XML based on attribute field and name - if you know one I can simplify this iteration and one or two other occurrences of similar logic.

Reply:
Hm, while a helper function is very often introduced to reuse code later, it is not the only reason it can be introduced. In this case, my goal is to make this function easier to read by replacing a simple operation with many lines of code with the operation itself. Developers only need to "read" this longer code if they believe there is a bug in it, otherwise they can continue on quickly when trying to understand the system. I do think your comment does something similar and if we remove the continue statements, the logic gets much shorter. Given the total length of this parent function, I think it's fine to leave the logic in.